edit/lint.js

```diff
@@ -0,0 +1,413 @@
+    template = ruleID => {
+      const rule = findCSSLintRule(ruleID);
+      return rule ? `<li><b>${makeLink(rule.url || url, rule.name)}</b><br>${rule.desc}</li>` : '';
+    };
+  } else {
+    header = t('issuesHelp', makeLink(url, 'stylelint'));
+    template = rule => `<li>${makeLink(url + rule, rule)}</li>`;
+  }
+  // to-do: change this to a generator
+  $$('#lint td[role="severity"]').forEach(el => {
+    const rule = el.dataset.rule;
+    if (!rules.includes(rule)) {
+      list += template(rule);
+      rules.push(rule);
+    }
+  });
+  return showHelp(t('issues'), header + list + '</ul>');
+}
+
+function showLinterErrorMessage(title, contents) {
+  messageBox({
+    title,
+    contents,
+    className: 'danger center lint-config',
+    buttons: [t('confirmOK')],
+  });
+}
+
+function showSavedMessage() {
+  $('#help-popup .saved-message').classList.add('show');
+  clearTimeout($('#help-popup .contents').timer);
+  $('#help-popup .contents').timer = setTimeout(() => {
+    // popup may be closed at this point
+    const msg = $('#help-popup .saved-message');
+    if (msg) {
+      msg.classList.remove('show');
+    }
+  }, 2000);
+}
+
+function checkLinter(linter = prefs.get('editor.linter')) {
+  linter = linter.toLowerCase();
+  if (prefs.get('editor.linter') !== linter) {
+    prefs.set('editor.linter', linter);
+  }
+  return linter;
+}
+
+function checkRules(linter, rules) {
+  const invalid = [];
+  const linterRules = linter === 'stylelint'
+    ? Object.keys(window.stylelint.rules)
+    : CSSLint.getRules().map(rule => rule.id);
+  Object.keys(rules).forEach(rule => {
+    if (!linterRules.includes(rule)) {
+      invalid.push(rule);
+    }
+  });
+  return invalid;
+}
+
+function setupLinterSettingsEvents(popup) {
+  $('.save', popup).addEventListener('click', event => {
+    event.preventDefault();
+    const linter = checkLinter(event.target.dataset.linter);
+    const json = tryJSONparse(popup.codebox.getValue());
+    if (json && json.rules) {
+      const invalid = checkRules(linter, json.rules);
+      if (invalid.length) {
+        return showLinterErrorMessage(
+          linter,
+          t('setLinterInvalidRuleError') + `<ul><li>${invalid.join('</li><li>')}</li></ul>`
+        );
+      }
+      if (linter === 'stylelint') {
+        setStylelintRules(json.rules);
+      } else {
+        setCSSLintRules(json.rules);
+      }
+      updateLinter(linter);
+      showSavedMessage();
+    } else {
+      showLinterErrorMessage(linter, t('setLinterError'));
+    }
+  });
+  $('.reset', popup).addEventListener('click', event => {
+    event.preventDefault();
+    const linter = checkLinter(event.target.dataset.linter);
+    let rules;
+    if (linter === 'stylelint') {
+      setStylelintRules();
+      rules = {rules: stylelintDefaultConfig.rules};
+    } else {
+      setCSSLintRules();
+      rules = {rules: csslintDefaultRuleset};
+    }
+    popup.codebox.setValue(JSON.stringify(rules, null, 2));
```

Comment:
The result of JSON.stringify for rules isn't easy to oversee at a glance, mainly because of `{"severity": "warning"}` occupying 3 lines.

Try tweaking it with something like `.replace(/,\n\s+\{\n\s+("severity":\s"\w+")\n\s+\}/gm, ', {$1}')`

Reply:
correction, `m` flag is not required in regexp.

Reply:
What do you think of putting it all on one line, except for the extended options. Something like this:

```js
{
  "at-rule-no-unknown": [true, {"severity": "warning"}],
  "block-no-empty": [true, {"severity": "warning"}],
  "color-no-invalid-hex": [true, {"severity": "warning"}],
  "declaration-block-no-duplicate-properties": [true,
  {
    "ignore": ["consecutive-duplicates-with-different-values"],
    "severity": "warning"
  }],
  "declaration-block-no-shorthand-property-overrides": [true, {"severity": "warning"}],
  "font-family-no-duplicate-names": [true, {"severity": "warning"}],
  "function-calc-no-unspaced-operator": [true, {"severity": "warning"}],
  "function-linear-gradient-no-nonstandard-direction": [true, {"severity": "warning"}],
  "keyframe-declaration-no-important": [true, {"severity": "warning"}],
  "media-feature-name-no-unknown": [true, {"severity": "warning"}],
  "no-empty-source": false,
  "no-extra-semicolons": [true, {"severity": "warning"}],
  "no-invalid-double-slash-comments": [true, {"severity": "warning"}],
  "property-no-unknown": [true, {"severity": "warning"}],
  "selector-pseudo-class-no-unknown": [true, {"severity": "warning"}],
  "selector-pseudo-element-no-unknown": [true, {"severity": "warning"}],
  "selector-type-no-unknown": [true, {"severity": "warning"}],
  "string-no-newline": [true, {"severity": "warning"}],
  "unit-no-unknown": [true, {"severity": "warning"}],
  "comment-no-empty": false,
  "declaration-block-no-redundant-longhand-properties": false,
  "shorthand-property-no-redundant-values": false
}
```

Reply:
It was my initial idea but then I decided not seeing true/false at a predictable position is rather confusing.